nixos/modules/services/web-apps/misskey.nix

```diff
@@ -0,0 +1,361 @@
+        };
+      };
+      redis = {
+        createLocally = lib.mkOption {
+          type = lib.types.bool;
+          default = false;
+          description = "Create and use a local Redis instance. Overrides `settings.redis.host`";
+        };
+        passwordFile = lib.mkOption {
+          type = lib.types.nullOr lib.types.path;
+          default = null;
+          description = "The path to a file containing the Redis password. Overrides `settings.redis.pass`.";
+        };
+      };
+      meilisearch = {
+        createLocally = lib.mkOption {
+          type = lib.types.bool;
+          default = false;
+          description = "Create and use a local Meilisearch instance. Overrides `settings.meilisearch.{host,port,ssl}`";
+        };
+        keyFile = lib.mkOption {
+          type = lib.types.nullOr lib.types.path;
+          default = null;
+          description = "The path to a file containing the Meilisearch API key. Overrides `settings.meilisearch.apiKey`.";
+        };
+      };
+      reverseProxy = {
+        enable = lib.mkEnableOption "a HTTP reverse proxy for Misskey";
+        webserver = lib.mkOption {
+          type = lib.types.attrTag {
+            nginx = lib.mkOption {
+              type = lib.types.submodule (import ../web-servers/nginx/vhost-options.nix);
+              default = { };
+              description = ''
+                Extra configuration for the nginx virtual host of Misskey.
+                Set to `{ }` to use the default configuration.
+              '';
+            };
+            caddy = lib.mkOption {
+              type = lib.types.submodule (import ../web-servers/caddy/vhost-options.nix { cfg = config.services.caddy; });
+              default = { };
+              description = ''
+                Extra configuration for the caddy virtual host of Misskey.
+                Set to `{ }` to use the default configuration.
+              '';
+            };
+          };
+          description = "The webserver to use as the reverse proxy.";
+        };
+      };
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    services.misskey.settings = {
+      db = (lib.optionalAttrs cfg.database.createLocally {
+        db = "misskey";
+        # Use unix socket instead of localhost to allow PostgreSQL peer authentication,
+        # required for `services.postgresql.ensureUsers`
+        host = "/var/run/postgresql";
+        port = config.services.postgresql.settings.port;
+        user = "misskey";
+        pass = null;
+      }) // (lib.optionalAttrs (cfg.database.passwordFile != null) {
+        pass = "@DATABASE_PASSWORD@";
+      });
+      redis = (lib.optionalAttrs cfg.redis.createLocally {
+        host = "localhost";
+      }) // (lib.optionalAttrs (cfg.redis.passwordFile != null) {
+        pass = "@REDIS_PASSWORD@";
+      });
+      meilisearch =
+        if (cfg.meilisearch.createLocally || (cfg.meilisearch.keyFile != null)) then
```

Comment:
This `if` expression breaks the ability to configure meilisearch via `settings.meilisearch`, resulting in `The option `services.misskey.settings.meilisearch` is defined both null and not null`.

The reason why I can't simply use `lib.optionalAttrs` here is that `settings.meilisearch = null` and `settings.meilisearch = { }` are parsed differently by Misskey. `settings.meilisearch` being `null` means Misskey won't try to use meilisearch, while `settings.meilisearch` being `{ }` results in an error.

I'll need to look into how to do this differently.